### PR TITLE
SLING-11736 support new "ensure nodes ..." statement

### PR DESCRIPTION
**Pull request by kwin**
Also fixes SLING-10463 by bailing out in case same name property exists

### CODE REVIEW

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java [hunk 1]
```diff
@@ -57,18 +59,33 @@ public void visitCreatePath(CreatePath cp) {
...
-                if (session.itemExists(fullPath)) {
-                    log.info("Path already exists, nothing to do (and not checking its primary type for now): {}", fullPath);
-                } else {
+                final Node node;
+                if (session.nodeExists(fullPath)) {
+                    log.info("Node at {} already exists, checking/adjusting its types", fullPath);
+                    node = session.getNode(fullPath);
+                    if (psd.getPrimaryType() != null && !node.getPrimaryNodeType().getName().equals(psd.getPrimaryType())) {
+                        log.info("Adjusting primary type of node {} to {}", fullPath, psd.getPrimaryType());
+                        node.setPrimaryType(psd.getPrimaryType());
```
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:68]: IMHO this is a fundamental change to the way repo-init deals with the 'create path' statement and i am not entirely sure if it wouldn't be better to use a dedicated syntax for enforcing the primary type.
personally, i would prefer if the primary type would have been set/adjusted from the very first version of repoinit. but i fear that changing it now will cause regressions as consumers of repoinit may have flawed type information with create-path statements that just work because a given node already exists.... changing the behavior now would result in these mistakes to become visible and thus failing the repository start up.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java [hunk 2]
```diff
@@ -57,18 +59,33 @@ public void visitCreatePath(CreatePath cp) {
...
-                    List<String> mixins = psd.getMixins();
-                    if (mixins != null) {
-                        log.info("Adding mixins {} to node {}", mixins, fullPath);
-                        for (String mixin : mixins) {
-                            node.addMixin(mixin);
-                        }
+                    node = addChildNode(parent, psd);
+                    
+                } else {
+                    throw new RepoInitException("There is a property with the name of the to be created node already at " + fullPath + ", therefore bailing out here as potentially not supported by the underlying JCR");
```
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:76]: i don't think repo-init should check this. JCR specification leaves it to the implementation if same-name property/node are allowed and IMHO repo-init should delegate that verification.
i might be mistaken but i vaguely remember that oak actually supports it.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java [hunk 3]
```diff
@@ -57,18 +59,33 @@ public void visitCreatePath(CreatePath cp) {
...
+                }
+                List<String> mixinsToAdd = new ArrayList<>(psd.getMixins());
+                // don't add mixins already set on the node
```
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:79]: what is the reason for omitting mixins that are already set?
also this changes looks unrelated to the SLING-11736 and i would suggest to make the change with a separate PR (if really needed)

**Review by anchela — CHANGES_REQUESTED** [review]
hi @kwin , while i agree that enforcing the primary type information would have been the better choice in the original design, i fear that changing it now will cause regressions. given that failures in repoinit will prevent the repository from starting altogether, i would prefer if enforcing type would be an optional feature that needs to be declared rather than changing the default.

**Comment by kwin** [comment]
> failures in repoinit will prevent the repository from starting altogether

We are in a predicament because without it the repository may not start either (and in fact already happened in https://github.com/adobe/aem-project-archetype/issues/997). 

In general the idea is that all statements written in repoinit are prerequisites, i.e. the repo should not start if it cannot apply them for whatever reason! Trying best effort is not good enough for repoinit.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java (hunk 3 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:79]: Well, not really unrelated because the idea is to set both mixins and primary type on existing nodes. The idea was to have some kind of (maybe premature) optimization here, and to allow for better logging (to distinguish between mixins already set and mixins added through repoinit).

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java (hunk 2 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:76]: you stated the opposite in https://issues.apache.org/jira/browse/SLING-10463.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java (hunk 1 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:68]: see https://github.com/apache/sling-org-apache-sling-jcr-repoinit/pull/41#issuecomment-1353043016

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java (hunk 3 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:79]: What issue do you see here?

Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:79]: no particular issue.... just seems unrelated to the original ticket and i am usually extra careful not to mix different issues.
if you wanted change the semantic of the create path to enforce all node types, you might as well have to remove mixins that are not declared but present... but that would definitely cause regressions.

**Comment by kwin** [comment]
@bdelacretaz I think your comment as the original inventor of repoinit would be helpful in this. The fundamental question really is: Best effort or fail? Currently we have a fail behaviour for all statements except for "create path" and "create principal acl" (https://github.com/apache/sling-org-apache-sling-jcr-repoinit/pull/14/files).

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java (hunk 3 above)
Comment (kwin) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:79]: Can you think of situations where additional mixins of existing nodes do harm?

File: src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java (hunk 2 above)
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:76]: yes, that can happen.... depending on whether i am looking at a problem from a oak point of view or from a dev caught up in an AEM escalation.

Comment (kwin) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/NodeVisitor.java:76]: that makes it hard to discuss and leads to frustration ;-)

**Comment by cziegeler** [comment]
I think we are past the point of being able to make breaking changes. It is not a question of best effort or fail - it is a question of a breaking change that will cause pain to our users. I've seen in other places how much pain it can cause if the code suddenly switches from best effort to fail, because now a lot of users run into the issue and projects that worked for years suddenly start to fail. We must not repeat this bad experience.

**Comment by kwin** [comment]
> We must not repeat this bad experience.

If developers cannot rely on the repo having the state as prescribed in repoinit there must be a lot of boilerplate code written by developerss how to it should deal with unexpected repository content. From my perspective this is the primary goal of repoinit (and executing this prior to any other code based on SlingRepository). If you cannot rely on that, that must be very explicitly stated in the documentation and kind of makes repoinit useless for most use cases (as then you can do as well do the changes manually in your code)

**Comment by cziegeler** [comment]
I have worked with users over the past year a lot where exactly such a change from best effort to fail caused so much pain - and while it seems logical to fix the code, users do not understand this. Their code worked yesterday, you changed something, now their code does not work anymore. And that is then the end of the discussion.
I think the best way out is to create new instructions with the right behaviour, deprecate the old instructions.

**Comment by kwin** [comment]
Ok, so what about a global "strict" flag (in addition to new instruction) set via OSGi configuration (for the default). For the new instruction I propose `create node` (with the same options as `create path`, as I consider the latter a poor name anyways, because it is misleading, as in JCR you are also creating a path when you add a new property).

**Comment by bdelacretaz** [comment]
>  I think the best way out is to create new instructions with the right behaviour, deprecate the old instructions.

I agree that repoinit should stay strictly backwards compatible, but for this specific case I think a _strict_ option on `create path` can do the job, without having to create completely new instructions (at least at the grammar level) and without requiring new configurations.

How about

    create path [strict] (nt:folder) /one(mixin nt:art)/step(mixin nt:dance)/two/steps

where `[strict]` is a new option, just for the `create path` statement, that activates the new behavior that this issues aims to implement?

That might cause execution of different code in the `jcr-repoinit` module, but at the grammar level it's a simple change, that IMO clearly expresses what's going on.

Doing it this means having to change existing repoinit statements, but I think it's safer than changing the behavior for existing code.

**Comment by kwin** [comment]
@bdelacretaz Let us please discuss this in https://github.com/apache/sling-org-apache-sling-repoinit-parser/pull/27. From an end user perspective deprecating one statement and introducing a new one with a more concise name is better than just adding the option strict to existing statements with fuzzy names like “path”

**Comment by bdelacretaz** [comment]
I suggest discussing on the dev list then, I think a more general discussion is useful on the topic of repoinit backwards compatibility and how we tackle similar cases in the future.

**Comment by kwin** [comment]
> I suggest discussing on the dev list then, I think a more general discussion is useful on the topic of repoinit backwards compatibility and how we tackle similar cases in the future.

Done in https://lists.apache.org/thread/7mvc2jvxltvrrntklzm1gzfmf0jy3mhc

**Review by anchela — APPROVED** [review]
@kwin , looks good to me.
i just have one more wish: the sling cp-2-featuremodel converter also relies on repoinit and we had troubles in the past when we expanded the parser API and forgot to adjust the converter. could you create a blocker ticket for the converter to make sure we check this before cutting a new release with the latest repoinit? thanks.
cc: @rombert , @karlpauls

**Comment by kwin** [comment]
> create a blocker ticket for the converter to make sure we check this before cutting a new release with the latest repoinit?

Done in https://issues.apache.org/jira/browse/SLING-11750.
